Automatically find externals

scheibelp opened the pull request:
Add a `spack external find` command which searches through `$PATH` and tries to create a `packages.yaml` of external packages. This focuses on finding build dependencies (currently only `cmake`).

To make a package discoverable with `spack external find` the package must define an `.executables` class attribute: a collection of executable names. If Spack finds these executables, it will create a `packages` config entry for them. If the package also defines `.determine_spec_details`, the executables will be passed to that function, which can generate a more-specific `packages.yaml` entry (including version, variants or whatever else the user decides to determine).

This currently outputs the relevant yaml configuration to the terminal and doesn't have an option to generate a `packages.yaml` file.

Closes #2507
Fixes #13198
Fixes #2116

alalazo commented:
I think the core proposal in #1795 might be of interest for a broader discussion. The basic idea is that a package that is detected as an external can be symlinked into a Spack owned directory. This would ensure that:
1. We don't introduce spurious dependencies by adding system folders as prefixes
2. We have a correct modelling of what needs to be detected for any external package that supports the detection API (if a critical component is missing, recipes are more likely to fail if we are out of system paths)

scheibelp commented:
> [This would ensure that] We have a correct modelling of what needs to be detected for any external package that supports the detection API (if a critical component is missing, recipes are more likely to fail if we are out of system paths)

I would like to give users the option to specify a subset of files (e.g. if a `cmake` exe is available on the system, I assume it is likely to work). Allowing users to specify more required files could be useful.

> [This would ensure that] We don't introduce spurious dependencies by adding system folders as prefixes

I think this can be treated as a separate issue: we already allow users to add external packages which are located in system folders (and in that case they enter the system folder as the prefix). Spack's logic already handles this case so it should be the same for automatically adding externals with system prefixes.

> The basic idea is that a package that is detected as an external can be symlinked into a Spack owned directory.

Overall I'm not sure of the benefit of this (vs. detecting and using system prefixes). I admit it *sounds* good to me, but I think we already need to handle the existence of system prefixes, so I'm not sure if separating the files into a different prefix is beneficial.

alalazo commented on lib/spack/spack/cmd/external.py:
```diff
@@ -0,0 +1,144 @@
+    exe_to_path = {}
+    # Reverse order of search directories so that an exe in the first PATH
+    # entry overrides later entries
+    for search_path in reversed(search_paths):
```
```suggestion
    for search_path in filter(os.path.exists, reversed(search_paths)):
```
The current code errors out if there's a path in `PATH` that does not exist

scheibelp replied:
Done

alalazo commented on lib/spack/spack/cmd/external.py:
```diff
@@ -0,0 +1,144 @@
+    # entry overrides later entries
+    for search_path in reversed(search_paths):
+        for exe in os.listdir(search_path):
+            exe_to_path[exe] = os.path.join(search_path, exe)
```
Would it be worthwhile to filter further with something simple like:
```python
def is_executable(path):
    return os.path.isfile(path) and os.access(path, os.X_OK)
```
?

scheibelp replied:
Done

alalazo commented on lib/spack/spack/cmd/external.py:
```diff
@@ -0,0 +1,144 @@
+                exe_pattern_to_pkgs[exe].append(pkg)
+
+    pkg_to_found_exes = defaultdict(set)
+    found_exe_to_pkgs = defaultdict(set)
```
A many to many relation seems appropriate here :+1:

alalazo commented on lib/spack/spack/cmd/external.py:
```diff
@@ -0,0 +1,144 @@
+
+            found_pkg_exes = pkg_to_found_exes[pkg]
+
+            if hasattr(pkg, 'determine_spec_details'):
+                spec_str = pkg.determine_spec_details(found_pkg_exes)
+            else:
+                spec_str = pkg.name
```
Does this assume a single spec? For compilers we might be dealing with cases like:
```console
$ gcc<Tab>
gcc             gcc-7           gcc-ar-4.8      gcc-ar-8        gcc-nm-5        gcc-nm-9        gcc-ranlib-6    
gcc-4.8         gcc-8           gcc-ar-5        gcc-ar-9        gcc-nm-6        gcc-ranlib      gcc-ranlib-7    
gcc-5           gcc-9           gcc-ar-6        gcc-nm          gcc-nm-7        gcc-ranlib-4.8  gcc-ranlib-8    
gcc-6           gcc-ar          gcc-ar-7        gcc-nm-4.8      gcc-nm-8        gcc-ranlib-5    gcc-ranlib-9   
```
so I'm wondering if that is covered here. What I would expect above is that multiple specs are deduced from the same package.

scheibelp replied:
This part doesn't but the logic above this needs to be adjusted

```
            if pkg.name in resolved_pkgs:
                 # The package was already considered as part of examining a
                 # different exe. It does not need to be reexamined.
                 continue
             resolved_pkgs.add(pkg.name)
```

it assumes we'll find one instance of any given package.

scheibelp commented:
(EDIT 4/21: the API for `determine_spec_details` has changed again since this comment)

To address https://github.com/spack/spack/pull/15158#discussion_r398001588 I had to update the method signature of `determine_spec_details`. The original signature:

`def determine_spec_details(exes)`

The new signature:

`def determine_spec_details(exe, all_files)`

The difference is that now this function is called with a single file as its first argument - this is the primary file to determine the spec for. Before, when this function was provided with a collection of files (possibly different versions of the same package), it wouldn't know which to return a spec for.

`all_files` is anticipated to be useful for packages which might want to examine multiple executables together (and later on, libraries/includes/config files) to decide what the spec is.

scheibelp commented:
@alalazo this is ready for another high-level look

scheibelp commented:
@alalazo @tgamblin based on our earlier conversation, I have updated the method signature of `determine_spec_details` again:

```
def determine_spec_details(prefix, exes_in_prefix):
  # now this can return one spec or several
```

This is called once for any pair of (prefix, package) - all the executables in the same prefix are grouped together in `exes_in_prefix`.

Note that without recording additional details for each spec, returning multiple specs for the same package and prefix won't be useful (e.g. it would need something like the compiler logic which can identify the individual executables associated with a specific version).

The following still applies:

* This can find different instances of a given package in different prefixes
* Any given spec can only be recorded as an external once (recording more than that wouldn't be meaningful)

alalazo commented on lib/spack/spack/cmd/external.py:
```diff
@@ -0,0 +1,194 @@
+    assumed there are two different instances of the executable.
+    """
+    path = os.getenv('PATH')
+    search_paths = list(p for p in path.split(os.pathsep) if os.path.isdir(p))
```
We can use functions already in other modules:
```suggestion
    hints = spack.util.environment.get_path('PATH')
    search_paths = llnl.util.filesystem..search_paths_for_executables(*hints)
```

alalazo commented on lib/spack/spack/cmd/external.py:
```diff
@@ -0,0 +1,194 @@
+        return [x]
+
+
+def _determine_base_dir(prefix):
+    assert os.path.isdir(prefix)
+    if os.path.basename(prefix) == 'bin':
+        return os.path.dirname(prefix)
```
This function returns `None` if the executable is not in a `bin` subdirectory? Shouldn't it return `prefix` instead?

scheibelp replied:
Right now I don't think Spack would know what to do with a prefix that didn't have the usual layout (e.g. with `bin/`, which is what `build_environment` adds). I'll add some logic to skip it along with a debug message in that case.

alalazo commented on lib/spack/spack/cmd/external.py:
```diff
@@ -0,0 +1,194 @@
+
+    pkg_to_found_exes = defaultdict(set)
+    for path, exe in system_path_to_exe.items():
+        for exe_pattern, pkgs in exe_pattern_to_pkgs.items():
+            if re.search(exe_pattern, exe):
```
This is just a reminder for later, but since we search within a loop it might be good to `re.compile` the pattern in advance.

alalazo commented on lib/spack/spack/cmd/external.py:
```diff
@@ -0,0 +1,194 @@
+            # for one prefix, but without additional details (e.g. about the
+            # naming scheme which differentiates them), the spec won't be
+            # usable.
+            if hasattr(pkg, 'determine_spec_details'):
+                specs = _convert_to_iterable(
+                    pkg.determine_spec_details(prefix, exes_in_prefix))
+            else:
+                # Note: packages which do not implement
+                # 'determine_spec_details' should use exact-match regular
+                # expressions
+                specs = [spack.spec.Spec(pkg.name)]
```
Just thinking loud about this, but I wonder if we should mandate for packages that implement this protocol something like `determine_version` and then have `determine_spec_details` act on executables that are grouped by prefix AND version.

The idea is that the regex pattern will be used to find matching executables, then:
1. We group by prefix
2. We group each prefix by version
3. We call `determine_spec_details(prefix, versions, exes)`

scheibelp replied:
To make sure I understand: is this to handle cases where multiple versions can be found in the same prefix?

I was originally thinking we could provide a utility function to start with an identifier string (like "gcc") and then return all the prefix/suffix combinations (and also possibly filter them based on whether they look like versions i.e. to avoid pulling in a file like `gccmakedep` and assuming it has a version `makedep`). The user could call this inside of their `determine_spec_details` implementation

I was thinking it would be easier to allow the user to inspect candidate files more closely before deciding that a particular version was present. It does create some extra work by requiring them to call the function. I think your suggestion may be better (since they can test the suggested exe after the fact).

My only thought is that it would be good to allow the user to disable this for some packages, possibly in the config and/or possibly at the package level.

alalazo replied:
> To make sure I understand: is this to handle cases where multiple versions can be found in the same prefix?

Yes. More in general I wonder if we should always require that a version is detected. 

> My only thought is that it would be good to allow the user to disable this for some packages, possibly in the config and/or possibly at the package level

Do you think it is possible to come up with a protocol similar to that of build-systems? There each build system has "phases" which usually have a default implementation that permit users to implement some callback to influence them. Taking as an example `Autotools` what most packages need to do is to implement `configure_args` and rely on the defaults of the base class for phases. If need be though, packages can completely override one or more phases and achieve more generality.

scheibelp replied:
> I think your suggestion may be better (since they can test the suggested exe after the fact).

I was thinking about your suggestion more and I'm actually not sure if it is any easier for the user to do things this way. Right now, the compiler logic groups executables by suffixes/prefixes that are extracted from the name (in make_compiler_list), but I’m wondering if this is necessary (for compilers or for detecting external packages):

If there are many instances of an installed package in a prefix, we will have to invoke each of the executables to determine what spec they are associated with, regardless of the naming scheme. Therefore, we can group executables by spec with or without being able to extract prefixes/suffixes from the naming scheme.

At least, this works well enough for versions. A naming scheme may provide more details than just the version, and it may be more difficult to learn these details by invoking an executable. I think a utility function could be provided so that a user could group all the executables by prefix/suffix with respect to an anchor string though in this case.

> I wonder if we should always require that a version is detected.

When examining packages with executables, the user should be able to verify that the output matches what they expect, so generally I agree that it should be possible to detect a version. I think that means that any package which defines `executables` must also define `determine_spec_details` (right now that’s optional) - do you agree?

> Do you think it is possible to come up with a protocol similar to that of build-systems? There each build system has "phases" which usually have a default implementation that permit users to implement some callback to influence them. 

I think that works well for build systems because they are established. My inclination is to provide utility functions to avoid constraining the user in case their package doesn't conform to our expectations

tgamblin replied:
>  I think that means that any package which defines executables must also define determine_spec_details (right now that’s optional) - do you agree?

I agree with this.  I think we should encourage packages that try to do this to provide more information.  We *could* define a default `determine_spec_details` that invokes with `--version` or something, but I'm hesitant to do that as there is such a wide range of behaviors that executables can have, and even if `--version` works, every command has a different syntax.  So yes, require it.

On specs vs. versions, etc.: I like the idea of detecting specs, then grouping similar specs, because it's more general than what the compiler logic currently does.  The compilers currently group by prefix, suffix, and version, but that is very specialized. The prefixes and suffixes are really for toolchains or variants -- e.g., what arch is the compiler for, or what was it built by (e.g., `-mp` for `macports`).  I think that stuff will generally come out as a version suffix or as a variant on the spec with this new logic.  It also simplifies the code -- the compiler code passes around a bunch of tuples in weird ways to various threaded routines.  This could generate specs first and consolidate the grouping logic in one place.

Only thing is we do not currently have a "general" way to group specs with similar attributes.  I guess there is `llnl.util.lang.index_by` -- it's used in some of the compiler commands. But then again, compilers only group things to try to match corresponding cc/cxx/fortran executables together.  We wouldn't need that yet.  Here, we are just saying "this prefix contains these specs", so if you have the specs do you need much else?

scheibelp replied:
> so if you have the specs do you need much else?

No (IMO), but my thought was there may be occasions where the `package.py`'s `determine_spec_details` implementation would have trouble determining the full spec, and may have to depend on the name. In that case, I think being able to parse out prefixes/suffixes would be useful, but not necessary for an initial cut.

alalazo replied:
> If there are many instances of an installed package in a prefix, we will have to invoke each of the executables to determine what spec they are associated with, regardless of the naming scheme. Therefore, we can group executables by spec with or without being able to extract prefixes/suffixes from the naming scheme.

I wonder if there are niche cases where this is not true. For compilers a common prefix/suffix usually indicates that executables belong to the same spec. In any case thinking about how this PR could evolve later I don't see us cornered in any way. The API in `determine_spec_details` is general enough to permit whatever refinement in its implementation.

As a point for later discussion, we might think if it makes sense to design different mixins depending on the type of package. Each mixin might provide a default implementation for `determine_spec_details` and require/permit to implement more specific methods with a narrower scope e.g. `determine_version` or whatever makes sense for that package type.

> I think that means that any package which defines executables must also define determine_spec_details (right now that’s optional) - do you agree?

+1 on this

> This could generate specs first and consolidate the grouping logic in one place.

+1 also to this

alalazo commented on var/spack/repos/builtin/packages/cmake/package.py:
```diff
@@ -13,6 +16,8 @@ class Cmake(Package):
     url      = 'https://github.com/Kitware/CMake/releases/download/v3.15.5/cmake-3.15.5.tar.gz'
     maintainers = ['chuckatkins']
 
+    executables = ['cmake']
```
I'll try tomorrow to play a bit with this package and see how easy it is to detect variants based on `ccmake` and other executables being found.

tgamblin commented on lib/spack/spack/cmd/external.py:
```diff
@@ -0,0 +1,204 @@
+    path_to_exe = {}
+    # Reverse order of search directories so that an exe in the first PATH
+    # entry overrides later entries
+    for search_path in reversed(search_paths):
```
This loop (and tests of executables) should be parallelized in a similar way to how compilers are parallelized (see the `ThreadPool` in `find_compilers` in `compilers/__init__.py`). Otherwise the search here will be extremely slow on systems like LC where there are lots of paths and most of the filesystem operations have NFS/Lustre/etc. latency.

scheibelp replied:
As far as finding the executables, the thread pool is used to invoke the executables vs. searching for them, and I don't think we can use threads to handle IO bottlenecks (i.e. the collection of the executables to test.

As far as parallelizing the running of the executables, right now we only run these tests for a a package that defines `executables` so we will be able to add optimization as things slow down (but for now only one package - i.e. CMake in this PR - has any invocations done for it).

tgamblin commented on lib/spack/spack/cmd/external.py:
```diff
@@ -0,0 +1,204 @@
+    """Generate config according to the packages.yaml schema for a single
+    package.
+
+    This does not generate the entire packages.yaml.
```
Docstring could use an example of the YAML that it generates.

tgamblin commented on lib/spack/spack/cmd/external.py:
```diff
@@ -0,0 +1,204 @@
+        system_path_to_exe = _get_system_executables()
+
+    exe_pattern_to_pkgs = defaultdict(list)
+    for pkg in repo.all_packages():
```
This will also be slow -- it's a good candidate for putting in the package cache (so we know which ones can be detected) but that can be a follow-on PR.

tgamblin commented on lib/spack/spack/cmd/external.py:
```diff
@@ -0,0 +1,204 @@
+    resolved_specs = {}  # spec -> exe found for the spec
+
+    # TODO: iterate through this in a predetermined order (e.g. by package
+    # name) to get repeatable results when there are conflicts.
```
can't we just sort the list to get this behavior?  Or do you think it would be better to resolve conflicts by `PATH` precedence?  What're the possible conflicts here?

scheibelp replied:
> What're the possible conflicts here?

If two different executables refer to the same spec, then Spack can only add one. I want it to add the same one each time `external find` is invoked. Sorting would get the desired behavior, although I think your point about PATH precedence is good too.

scheibelp replied:
I have updated and moved the TODO. It's not addressed yet but I'm not sure if that's essential for initial merge.

tgamblin commented on lib/spack/spack/cmd/external.py:
```diff
@@ -0,0 +1,204 @@
+                if not pkg_prefix:
+                    tty.debug("{0} does not end with a 'bin/' directory: it"
+                              " cannot be added as a Spack package"
+                              .format(prefix))
```
Is this 100% true?  There are packages that call `env.prepend_path()` in `setup_run_environment()`.  Could we use that information to allow other directories?  e.g., if we see something that starts with the `prefix` in the `EnvModifications` for `PATH` after calling `setup_run_environment()`, we know that the package can have executables in that directory.  I don't think this is necessary for the first cut, but it's a way we could easily consider packages that don't exclusively use `bin`.

scheibelp replied:
Based on your suggestion, you are right that this package could strive to see whether it can make some use out of the prefix; generally (unless the package is exclusively provided externally) I think the `package.py` implementation would expect a typical Spack layout. There may be a couple packages that accommodate this now, and the `setup_run_environment` function could be updated to be more flexible.

IMO this could be added as a TODO and addressed later.

tgamblin replied:
You could easily add that later by adding match groups to the regex, and that’s more flexible anyway

tgamblin commented on lib/spack/spack/cmd/external.py:
```diff
@@ -0,0 +1,204 @@
+                    ExternalPackageEntry(spec=spec, base_dir=pkg_prefix))
+            else:
+                tty.debug(
+                    'The following executables in {0} were decidedly not'
```
"decidedly" seems like overkill here, but I like the way it reads :).

alalazo replied:
@scheibelp I don't think I understand how this debug information is meant to be triggered. It's executed after the:
```python
for spec in specs:
    ...
```
loop and I don't see any `break` statement in the logic to avoid it under certain conditions. If I turn debug information on:
```console
$ spack -d external find cmake 
==> [2020-04-30-22:32:33.309250, 28329] Imported external from built-in commands
==> [2020-04-30-22:32:33.310512, 28329] Imported external from built-in commands
==> [2020-04-30-22:32:33.311191, 28329] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/defaults/repos.yaml
==> [2020-04-30-22:32:33.313666, 28329] Reading config file /home/culpo/.spack/linux/repos.yaml
==> [2020-04-30-22:32:33.411763, 28329] '/usr/bin/cmake' '--version'
==> [2020-04-30-22:32:33.416848, 28329] The following executables in /usr/bin were decidedly notpart of the package cmake: /usr/bin/ccmake, /usr/bin/cmake
==> [2020-04-30-22:32:33.417374, 28329] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml
==> [2020-04-30-22:32:33.432310, 28329] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/packages.yaml
```
The debug information is triggered even though:
```
$ cat /home/culpo/.spack/linux/packages.yaml
packages:
  cmake:
    paths:
      cmake@3.17.2: /usr
    version: []
    target: []
    compiler: []
    buildable: true
    modules: {}
    providers: {}
```
I think this part should be something like:
```python
if not specs:
   tty.debug('...')
```
before the loop instead of an `else` clause after it.

scheibelp replied:
I thought I was being smart but definitely got that wrong. Thanks!

scheibelp commented:
> As long as the command works and generates some information in packages.yaml

This requires discussion: should we dump this into the user's existing config? If so, and if our spec conflicts with a preexisting spec, should we overwrite it or should we skip it?

I think a sensible approach would be to write into the user's config but to avoid adding any specs that have already been defined.

> My main concerns is that exe detection will be potentially quite slow on our machines,

I don't think this will be initially problematic (especially for finding build deps). Once we move to detecting compiler versions using this logic then it will definitely be needed.

tgamblin commented:
> This requires discussion: should we dump this into the user's existing config? If so, and if our spec conflicts with a preexisting spec, should we overwrite it or should we skip it?

In current spack the policy is to write to the "default write scope", so we should do that, but that'll obviously change with #15939 later.  I think that's #15939's job to change.

> I think a sensible approach would be to write into the user's config but to avoid adding any specs that have already been defined.

An option for this would be nice (`-f`?)  Then again, `spack compiler find` does not have a `-f` and it just tells you what's been added.  Staying consistent with that seems fine for now.

> My main concerns is that exe detection will be potentially quite slow on our machines,
> > I don't think this will be initially problematic (especially for finding build deps). Once we move to detecting compiler versions using this logic then it will definitely be needed.

Ok fair enough -- let's see how it does, but it eventually does need to be fast.  Have you tried it on a cluster?

alalazo commented on lib/spack/spack/cmd/external.py:
```diff
@@ -0,0 +1,204 @@
+    sp = subparser.add_subparsers(
+        metavar='SUBCOMMAND', dest='external_command')
+
+    sp.add_parser('find', help=external_find.__doc__)
```
Just an idea, maybe for another PR, but I think in the long run we should enable users to require detection only for certain packages e.g.:
```console
$ spack external find cmake gcc
```
As more packages implement detection it might be the case that people want to register a selected subset of them.

tgamblin replied:
I like this — it would be pretty easy to do here wouldn’t it?

scheibelp commented:
@tgamblin @alalazo I have addressed or responded to review comments. All comments where I didn't change the code yet in response I have left expanded. Also, there is now some preliminary testing (which I plan to expand on tomorrow).

> An option for this would be nice (-f?) Then again, spack compiler find does not have a -f and it just tells you what's been added. Staying consistent with that seems fine for now.

I think we should be careful to avoid overwriting user-generated config. As of now the logic only inserts external entries if there is no existing entry for the spec (and uses `config._merge_yaml` in part to achieve this). I think it might be useful to distinguish between manual and automatic entries so that we could do a force update without overwriting user config. Another option is to use a different scope.

alalazo commented on lib/spack/spack/cmd/external.py:
```diff
@@ -0,0 +1,238 @@
+    if args.packages:
+        packages_to_check = list(spack.repo.get(pkg) for pkg in args.packages)
+    else:
+        packages_to_check = spack.repo.all_packages()
```
I think this should read:
```suggestion
        packages_to_check = spack.repo.path.all_packages()
```
What I get with the current version is:
```console
$ spack external find
==> Error: module 'spack.repo' has no attribute 'all_packages'
```

alalazo commented:
> I think it might be useful to distinguish between manual and automatic entries so that we could do a force update without overwriting user config. Another option is to use a different scope.

Having an additional scope at low priority with all the config files that might be generated automatically by Spack seems a good idea to me. Users can always override it with other scopes and we can regenerate the files without worrying too much that some user defined config is lost.

alalazo commented on lib/spack/spack/cmd/external.py:
```diff
@@ -0,0 +1,238 @@
+    else:
+        packages_to_check = spack.repo.all_packages()
+
+    pkg_to_entries = _get_external_packages(packages_to_check)
+    _update_pkg_config(pkg_to_entries)
```
We probably need to implement some consistency check of the detected specs. What I get right now in the YAML is:
```yaml
packages:
  cmake:
    paths:
      cmake@[Version('3.17.2')]: /usr
    version: []
    target: []
    compiler: []
    buildable: true
    modules: {}
    providers: {}
```
while I think we want:
```yaml
packages:
  cmake:
    paths:
      cmake@3.17.2: /usr
    ...
```
The issue is in the implementation of `CMake.determine_spec_details`, but I think we should avoid that something like that results in the generation of an invalid YAML.

scheibelp replied:
I added a `_spec_is_valid` function which would catch this issue. In general I think we'll have to make the Spec interface easier to use when constructing new specs. Even when that is updated, I think this function will be useful to detect issues (e.g. with the approach I originally took in the CMake package).

alalazo commented on var/spack/repos/builtin/packages/cmake/package.py:
```diff
@@ -141,6 +146,25 @@ class Cmake(Package):
+            match = re.search(r'version\s+(\S+)', output)
+            if match:
+                version_str = match.group(1)
+            spec.versions = [Version(version_str)]
```
I think this should read:
```suggestion
            spec.versions = Version(version_str)
```
for the version to be seen as concrete.

scheibelp replied:
This is strange: I generally like to build up a Spec object and then stringify it. When I did as you suggest using a nonstandard version:

`spec.versions = Version('1.foo')`

I got a strange error:

```
>       assert single_entry.spec == Spec('cmake@1.foo')

lib/spack/spack/test/cmd/external.py:47: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
lib/spack/llnl/util/lang.py:228: in <lambda>
    (s is o) or (o is not None and s._cmp_key() == o._cmp_key()))
lib/spack/spack/version.py:91: in coercing_method
    ca, cb = coerce_versions(a, b)
lib/spack/spack/version.py:68: in coerce_versions
    check_type(tb)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

t = <type 'int'>

    def check_type(t):
        if t not in order:
>           raise TypeError("coerce_versions cannot be called on %s" % t)
E           TypeError: coerce_versions cannot be called on <type 'int'>
```

So I think there might be an issue for how `Spec.__str__` works. IMO either mine or your expression should result in a valid string. I don't think we technically have an API for constructing Specs manually, so it might be better to suggest that users construct strings that represent Specs instead of the Specs themselves.

alalazo replied:
> I don't think we technically have an API for constructing Specs manually, so it might be better to suggest that users construct strings that represent Specs instead of the Specs themselves.

:100: on that. I think packagers would be in general more accustomed to work with spec strings that to construct a Spec object using the internal API.

alalazo commented on var/spack/repos/builtin/packages/cmake/package.py:
```diff
@@ -141,6 +146,25 @@ class Cmake(Package):
 
     phases = ['bootstrap', 'build', 'install']
 
+    @staticmethod
```
Is there a reason to ask for a `staticmethod` rather than a `classmethod`?

scheibelp replied:
I think each class could choose either. For class methods the class is provided as an additional initial argument. If the method implementation doesn't refer to class details, then it can use a `staticmethod`. Generally though classes probably will refer to such details though (e.g. the regexes used to find executables) so to set an example, I will change it to a `classmethod`.

scheibelp commented:
All new comments are addressed and new tests added. This includes:

* Added a `--not-buildable` option which automatically marks packages with collected externals as non-buildable
* Spec consistency checking (ensuring that constructed spec converts between Spec and string sensibly)
* Tests for the command (previously only the logic underlying it was tested) as well as config merging
* The mock repo includes a new test package which defines `determine_spec_details` (many of the tests are still implemented in terms of the "real" Spack CMake package - they could be converted if desired).
* Test for calling `spack external find` with no additional arguments (in which case all packages that define `determine_spec_details` are checked
* (EDIT: forgot to add) Define `determine_spec_details` for the Automake package (so now Spack can find externals for CMake and Automake)

scheibelp commented:
@tgamblin @alalazo this is ready for another look

scheibelp commented:
I should add a section on things that are not yet addressed but I presume they could be handled with a later PR:

* Parallelize the execution of `determine_spec_details` across different packages: https://github.com/spack/spack/pull/15158#discussion_r416806219
* Ensure repeatable results in the case of conflicts: https://github.com/spack/spack/pull/15158#discussion_r416982573
* Handle external packages with atypical layouts (i.e. not `<prefix>/`, `<prefix>/bin`, `<prefix/lib>`, etc.): https://github.com/spack/spack/pull/15158#discussion_r416817760
* (EDIT: moved from https://github.com/spack/spack/pull/15158#issuecomment-621526152) Add a new scope to register automatically found externals at (at the moment, this may not be critical as I have added logic/testing to ensure that existing spec entries are not overwritten)
* (done now) ~Use `re.compile` on all regexes (actually that is simple enough to do here but isn't done yet): https://github.com/spack/spack/pull/15158#discussion_r413218689~

All other concerns are addressed at this point.

alalazo commented on lib/spack/spack/cmd/external.py:
```diff
@@ -0,0 +1,271 @@
+    groups = defaultdict(set)
+    for p in paths:
+        groups[os.path.dirname(p)].add(p)
+    return groups.items()
```
I think we can return:
```python
result = sorted([(prefix, sorted(s)) for prefix, s in groups.items()])
```
to solve the TODO below

scheibelp replied:
That would do it, although the TODO I have in the code mentions that it might be better to do it based on order of prefixes in `PATH` (so the results are repeatable and follow an intuitive rule).

alalazo commented on var/spack/repos/builtin/packages/cmake/package.py:
```diff
@@ -141,6 +146,22 @@ class Cmake(Package):
+            match = re.search(r'cmake.*version\s+(\S+)', output)
+            if match:
+                version_str = match.group(1)
+                return Spec('cmake@{0}'.format(version_str))
```
Maybe we can just:
```suggestion
                return 'cmake@{0}'.format(version_str)
```
and move the logic that parses the string into Specs in the command, as you suggested in a comment above.

scheibelp commented:
> Can you please add documentation before we merge it?

@tgamblin I have added documentation in  https://github.com/spack/spack/pull/15158/commits/8af08724cf80010a2cd4fcf987b287c0032659cd

tgamblin commented on lib/spack/docs/build_settings.rst:
```diff
@@ -158,6 +158,37 @@ Spack can then use any of the listed external implementations of MPI
+Specific limitations include:
+
+* A package must define ``executables`` and ``determine_spec_details``
+  for Spack to locate instances of that package.
```
This needs further explanation in the packaging guide -- what's the signature for the method, what should `executables` contain, and how is finding done based on those two things?  Can you add that?

tgamblin commented on lib/spack/docs/build_settings.rst:
```diff
@@ -158,6 +158,37 @@ Spack can then use any of the listed external implementations of MPI
+* A package must define ``executables`` and ``determine_spec_details``
+  for Spack to locate instances of that package.
+* This is currently intended to find build dependencies rather than
+  library packages.
```
Probably should say why -- because it can only find executable programs right now.

tgamblin commented on lib/spack/docs/build_settings.rst:
```diff
@@ -158,6 +158,37 @@ Spack can then use any of the listed external implementations of MPI
+  can help locate all external entries).
+* Currently this logic is focused on examining ``PATH`` and does not
+  search through modules (although it should find the package if a
+  module is loaded for it).
```
I think this is related to the second bullet above and should be combined.

adamjstewart commented:
It seems like `spack external find` can only find the first instance of an external package? What if you have multiple copies of a package like Python installed and you want to pick up both Python 2 and 3?

adamjstewart commented:
Also, the `prefix` Spack chooses is non-deterministic. Sometimes I get my Spack-installed Python 3, sometimes I get my system Python 2, and sometimes I get `spack-python`.

P.S. I'm working on adding `spack external find python` support

adamjstewart commented:
Ah, now that I have it working it seems to run this function for each unique prefix. What about when both Python 2 and 3 are installed to the same directory?

alalazo commented:
> What about when both Python 2 and 3 are installed to the same directory?

@adamjstewart You might want to look at #16526  which builds on this PR and implements detection logic for `gcc`.

adamjstewart commented:
Also, this PR suffers from the same issue as #12313, it reorganizes my `packages.yaml`, including existing entries.

adamjstewart commented:
Btw, I got bored, so I'm going to add `spack external find` support for common build dependencies, like:
```console
$ egrep -ho 'depends_on\(.(\w+).*type=.build' */package.py | cut -c 13- | egrep -o '^[A-Za-z0-9-]+' | sort | uniq -c | sort -n
...
  38 gmake
  41 flex
  53 bison
  54 py-cython
  61 python
  79 xproto
 155 m4
 181 util-macros
 187 automake
 190 autoconf
 195 libtool
 286 cmake
 347 pkgconfig
 719 py-setuptools
```
